src/SapNwRfc/SapFunctionMetadata.cs

```diff
@@ -0,0 +1,115 @@
+using System;
+using SapNwRfc.Internal.Interop;
+
+namespace SapNwRfc
+{
+    /// <summary>
+    /// Represents SAP RFC function metadata.
+    /// </summary>
+    public sealed class SapFunctionMetadata : ISapFunctionMetadata
+    {
+        private readonly RfcInterop _interop;
+        private readonly IntPtr _functionDescHandle;
+
+        internal SapFunctionMetadata(RfcInterop interop, IntPtr functionDescHandle)
+        {
+            _interop = interop;
+            _functionDescHandle = functionDescHandle;
+        }
+
+        /// <inheritdoc cref="ISapFunctionMetadata"/>
+        public string GetName()
+        {
+            RfcResultCode resultCode = _interop.GetFunctionName(
+                rfcHandle: _functionDescHandle,
+                funcName: out string funcName,
+                errorInfo: out RfcErrorInfo errorInfo);
+
+            errorInfo.ThrowOnError();
+
+            return funcName;
+        }
+
+        /// <inheritdoc cref="ISapFunctionMetadata"/>
+        public ISapParameterMetadata GetParameterByIndex(uint index)
+        {
+            RfcResultCode resultCode = _interop.GetParameterDescByIndex(
+                funcDesc: _functionDescHandle,
+                index: index,
+                paramDesc: out RfcParameterDescription paramDesc,
+                errorInfo: out RfcErrorInfo errorInfo);
+
+            errorInfo.ThrowOnError();
+
+            return new SapParameterMetadata(_interop, paramDesc);
+        }
+
+        /// <inheritdoc cref="ISapFunctionMetadata"/>
+        public ISapParameterMetadata GetParameterByName(string name)
+        {
+            RfcResultCode resultCode = _interop.GetParameterDescByName(
+                funcDesc: _functionDescHandle,
+                name: name,
+                paramDesc: out RfcParameterDescription paramDesc,
+                errorInfo: out RfcErrorInfo errorInfo);
+
+            errorInfo.ThrowOnError();
+
+            return new SapParameterMetadata(_interop, paramDesc);
+        }
+
+        /// <inheritdoc cref="ISapFunctionMetadata"/>
+        public uint GetParameterCount()
```

Comment:
I wonder if we need to expose `GetParameterCount` / `GetParameterDescByIndex` and `GetExceptionCount` / `GetExceptionByIndex` and not just a `IEnumerable<ISapParameterMetadata> Parameters` and `IEnumerable<ISapExceptionMetadata> Exceptions` property. Would also be nice if we could cache the result, so multiple enumerations of the properties don't invoke the low-level API functions again.

Is there any use for looking up a parameter or exception by name?

Reply:
We could even expose them as a `IReadOnlyList<ISapParameterMetadata> Parameters` so you can still access by index and get the count (as int). Or as a custom `IReadOnlyListDictionary` interface which also implements `IReadOnlyDictionary` so you could also access them by `name`,

---

Looking up parameters / exceptions by name would be needed for accessing them dynamically, e.g.:
```
var result = fn.Invoke<dynamic>();
var value = result.PARAMETER_NAME;
```

But this could also be accomplished by looking it up in the `IEnumerable<ISapParameterMetadata> Parameters`.